native/src/main/java/io/ballerina/stdlib/xmldata/XmlToJson.java

```diff
@@ -154,73 +152,88 @@ private static Object convertElement(BXmlItem xmlItem, String attributePrefix,
-        if (attributeManager.getMap().isEmpty()) {
-            tempAttributeMap = collectAttributesAndNamespaces(xmlItem, preserveNamespaces);
-            attributeManager.initializeMap(tempAttributeMap);
+        Map<String, String> nsPrefixMap = getNamespacePrefixes(attributeMap);
+        if (prefix != null && preserveNamespaces && parentAttributeMap != null) {
+            for (Map.Entry<BString, BString> entry : attributeMap.entrySet()) {
+                BString value = entry.getValue();
+                if (!isNamespacePrefixEntry(entry) ||
+                        isBelongingToElement(parentAttributeMap, entry.getKey(), value)) {
+                    String key = attributePrefix + getKey(entry, nsPrefixMap, preserveNamespaces);
+                    putAsFieldTypes(mapData, key, value.getValue(), type);
+                }
+            }
         } else {
-            LinkedHashMap<String, String> newAttributeMap = collectAttributesAndNamespaces(xmlItem, preserveNamespaces);
-            LinkedHashMap<String, String> attributeMap = attributeManager.getMap();
-            for (Map.Entry<String, String> entrySet : newAttributeMap.entrySet()) {
-                String key = entrySet.getKey();
-                String value = entrySet.getValue();
-                if (attributeMap.get(key) == null || !attributeMap.get(key).equals(value)) {
-                    attributeManager.setValue(key, value);
-                    tempAttributeMap.put(key, value);
+            for (Map.Entry<BString, BString> entry : attributeMap.entrySet()) {
+                String key = getKey(entry, nsPrefixMap, preserveNamespaces);
+                if (key != null) {
+                    putAsFieldTypes(mapData, attributePrefix + key, entry.getValue().getValue(), type);
                 }
             }
         }
-        addAttributes(mapData, attributePrefix, tempAttributeMap, type, uniqueKey);
     }
 
-    private static void addAttributes(BMap<BString, Object> rootNode, String attributePrefix,
-                                      LinkedHashMap<String, String> attributeMap, Type type, String uniqueKey) {
-        for (Map.Entry<String, String> entry : attributeMap.entrySet()) {
-            putAsFieldTypes(rootNode, attributePrefix + entry.getKey(), entry.getValue(), type, uniqueKey);
-        }
+    private static boolean isBelongingToElement(BMap<BString, BString> parentAttributeMap, BString key,
+                                                       BString value) {
+        return !(parentAttributeMap.containsKey(key) &&
+                parentAttributeMap.get(key).getValue().equals(value.getValue()));
     }
 
-    private static void putAsFieldTypes(BMap<BString, Object> map, String key, String value, Type type,
-                                        String uniqueKey) {
+    private static void putAsFieldTypes(BMap<BString, Object> map, String key, String value, Type type) {
         if (type != null) {
```

Comment:
We don't need the `null` check here since the `instance of` checks in `if` and `else if` will return false (https://stackoverflow.com/questions/2950319/is-null-check-needed-before-calling-instanceof).

Reply:
This is a common function of toJson and toRecord.  So, this null check is used to identify which function is called. If that is toJson, the type is null.

Reply:
If it is `toJson`, then the `type` is `null` and the following will return `null` right?
```
private static Type getFieldType(String fieldName, Type type) {
        if (type instanceof RecordType) {
            if (((RecordType) type).getFields().get(fieldName) != null) {
                return ((RecordType) type).getFields().get(fieldName).getFieldType();
            }
        } else if (type instanceof ArrayType) {
            Type filedType = ((ArrayType) type).getElementType();
            if (filedType instanceof RecordType) {
                Map<String, Field> fileds = ((RecordType) filedType).getFields();
                if (fileds.get(fieldName) != null) {
                    return fileds.get(fieldName).getFieldType();
                }
            }
            return filedType;
        }
        return type;
    }
```

Reply:
Yes, we can do it. 

 You put the comment in the wrong place. So I explained that function.

Reply:
Removed [b9ec199](https://github.com/ballerina-platform/module-ballerina-xmldata/pull/422/commits/b9ec199a40122337cd9b30b12cadab5a325778de)